sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala

```diff
@@ -167,6 +173,229 @@ final class DataFrameWriterV2[T] private[sql](table: String, ds: Dataset[T])
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @return a new `WhenNotMatchedBySource` object.
+   */
+  def whenNotMatchedBySource(): WhenNotMatchedBySource[T] = {
+    new WhenNotMatchedBySource[T](this, None)
+  }
+
+  /**
+   * Initialize a `WhenNotMatchedBySource` object with a condition.
+   *
+   * This `WhenNotMatchedBySource` action will be executed if and only if the specified `condition`
+   * is satisfied.
+   *
+   * This `WhenNotMatchedBySource` can be followed by one of the following merge actions:
+   *   - `updateAll`: Update all the target table fields with source dataset fields.
+   *   - `update(Map)`: Update all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `insertAll`: Insert all the target table with source dataset records.
+   *   - `insert(Map)`: Insert all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @param condition a `Column` representing the condition to be evaluated for the action.
+   * @return a new `WhenNotMatchedBySource` object configured with the specified condition.
+   */
+  def whenNotMatchedBySource(condition: Column): WhenNotMatchedBySource[T] = {
+    new WhenNotMatchedBySource[T](this, Some(condition.expr))
+  }
+
+  /**
+   * Initialize a `WhenNotMatchedBySource` object with a condition.
+   *
+   * This `WhenNotMatchedBySource` action will be executed if and only if the specified `condition`
+   * is satisfied. The condition is represented as a `String` and internally converted
+   * to a `Column`.
+   *
+   * This `WhenNotMatchedBySource` can be followed by one of the following merge actions:
+   *   - `updateAll`: Update all the target table fields with source dataset fields.
+   *   - `update(Map)`: Update all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `insertAll`: Insert all the target table with source dataset records.
+   *   - `insert(Map)`: Insert all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @param condition a `String` representing the condition to be evaluated for the action.
+   * @return a new `WhenNotMatchedBySource` object configured with the specified condition.
+   */
+  def whenNotMatchedBySource(condition: String): WhenNotMatchedBySource[T] = {
+    whenNotMatchedBySource(Column(condition))
+  }
+
+  /**
+   * Executes the merge operation.
+   */
+  def merge(): Unit = {
+    val merge = MergeIntoTable(
+      UnresolvedRelation(tableName),
+      logicalPlan,
+      on.get.expr,
+      matchedActions,
+      notMatchedActions,
+      notMatchedBySourceActions)
```

Comment:
Do we need some sanity checks for these parameters to `MergeIntoTable`? I.e., `matchedActions`, `notMatchedActions`, `notMatchedBySourceActions`? For example, can they all empty etc.

Reply:
Checks are added